File: web/lib/encoding.ts
// TypeScript port of trainer/chessai/encoding.py — must match it EXACTLY.
// We only need the forward direction in the browser:
//   * boardToPlanes(fen)  -> Float32Array (18*8*8), side-to-move perspective
//   * moveToIndex(...)    -> policy index for a legal move
// Parity is verified against the Python implementation in scripts/parity_check.

export const N_PLANES = 18;
export const N_MOVES = 64 * 73; // 4672

// piece char -> plane offset within a color's 6 planes
const PIECE_IDX: Record<string, number> = { p: 0, n: 1, b: 2, r: 3, q: 4, k: 5 };

// 8 compass directions (fileDelta, rankDelta) — order matters
const DIRECTIONS: [number, number][] = [
  [0, 1],   // N
  [1, 1],   // NE
  [1, 0],   // E
  [1, -1],  // SE
  [0, -1],  // S
  [-1, -1], // SW
  [-1, 0],  // W
  [-1, 1],  // NW
];

const KNIGHT_DELTAS: [number, number][] = [
  [1, 2], [2, 1], [2, -1], [1, -2],
  [-1, -2], [-2, -1], [-2, 1], [-1, 2],
];

const UNDERPROMO = ["n", "b", "r"]; // knight, bishop, rook

export interface ParsedFen {
  pieces: ({ type: string; color: "w" | "b" } | null)[]; // index 0..63, a1=0
  turn: "w" | "b";
  castling: string;
  epSquare: number; // 0..63 or -1
  halfmove: number;
}

// square index in python-chess convention: a1=0, b1=1, ... h8=63
function fileOf(sq: number): number { return sq & 7; }
function rankOf(sq: number): number { return sq >> 3; }
function mirrorSquare(sq: number): number { return sq ^ 56; } // vertical flip

export function squareToIndex(alg: string): number {
  // 'e2' -> file*1 + rank*8
  const file = alg.charCodeAt(0) - 97; // 'a'
  const rank = alg.charCodeAt(1) - 49; // '1'
  return rank * 8 + file;
}

export function parseFen(fen: string): ParsedFen {
  const parts = fen.trim().split(/\s+/);
  const [placement, turn, castling, ep, half] = parts;
  const pieces: ParsedFen["pieces"] = new Array(64).fill(null);
  const rows = placement.split("/"); // row 0 = rank 8
  for (let r = 0; r < 8; r++) {
    const actualRank = 7 - r;
    let file = 0;
    for (const ch of rows[r]) {
      if (ch >= "1" && ch <= "8") {
        file += parseInt(ch, 10);
      } else {
        const color = ch === ch.toUpperCase() ? "w" : "b";
        const sq = actualRank * 8 + file;
        pieces[sq] = { type: ch.toLowerCase(), color };
        file += 1;
      }
    }
  }
  return {
    pieces,
    turn: turn === "w" ? "w" : "b",
    castling: castling || "-",
    epSquare: ep && ep !== "-" ? squareToIndex(ep) : -1,
    halfmove: half ? parseInt(half, 10) : 0,
  };
}

export function boardToPlanes(fen: string): Float32Array {
  const p = parseFen(fen);
  const planes = new Float32Array(N_PLANES * 64);
  const stm = p.turn;
  const mirror = stm === "b";

  const set = (plane: number, rank: number, file: number, v: number) => {
    planes[plane * 64 + rank * 8 + file] = v;
  };

  for (let sq = 0; sq < 64; sq++) {
    const piece = p.pieces[sq];
    if (!piece) continue;
    const idxSq = mirror ? mirrorSquare(sq) : sq;
    const rank = rankOf(idxSq);
    const file = fileOf(idxSq);
    const own = piece.color === stm;
    const base = own ? 0 : 6;
    const plane = base + PIECE_IDX[piece.type];
    set(plane, rank, file, 1.0);
  }

  // castling rights from side-to-move perspective
  const wK = p.castling.includes("K");
  const wQ = p.castling.includes("Q");
  const bK = p.castling.includes("k");
  const bQ = p.castling.includes("q");
  const ownK = stm === "w" ? wK : bK;
  const ownQ = stm === "w" ? wQ : bQ;
  const oppK = stm === "w" ? bK : wK;
  const oppQ = stm === "w" ? bQ : wQ;
  const fillPlane = (plane: number) => {
    for (let i = 0; i < 64; i++) planes[plane * 64 + i] = 1.0;
  };
  if (ownK) fillPlane(12);
  if (ownQ) fillPlane(13);
  if (oppK) fillPlane(14);
  if (oppQ) fillPlane(15);

  // en passant file
  if (p.epSquare >= 0) {
    const ep = mirror ? mirrorSquare(p.epSquare) : p.epSquare;
    const file = fileOf(ep);
    for (let rank = 0; rank < 8; rank++) set(16, rank, file, 1.0);
  }

  // fifty-move clock
  const hv = Math.min(p.halfmove, 100) / 100.0;
  for (let i = 0; i < 64; i++) planes[17 * 64 + i] = hv;

  return planes;
}

function deltaToPlane(df: number, dr: number, promotion: string | null): number {
  if (promotion && promotion !== "q") {
    const pieceIdx = UNDERPROMO.indexOf(promotion);
    return 64 + pieceIdx * 3 + (df + 1);
  }
  // knight
  for (let i = 0; i < KNIGHT_DELTAS.length; i++) {
    if (KNIGHT_DELTAS[i][0] === df && KNIGHT_DELTAS[i][1] === dr) return 56 + i;
  }
  // queen-like
  const dist = Math.max(Math.abs(df), Math.abs(dr));
  const sf = df === 0 ? 0 : df / Math.abs(df);
  const sr = dr === 0 ? 0 : dr / Math.abs(dr);
  let dirIdx = -1;
  for (let i = 0; i < DIRECTIONS.length; i++) {
    if (DIRECTIONS[i][0] === sf && DIRECTIONS[i][1] === sr) { dirIdx = i; break; }
  }
  return dirIdx * 7 + (dist - 1);
}

// from/to are algebraic squares ('e2'), promotion is 'q'|'n'|'b'|'r'|null
export function moveToIndex(
  from: string, to: string, promotion: string | null, turn: "w" | "b"
): number {
  let fromSq = squareToIndex(from);
  let toSq = squareToIndex(to);
  if (turn === "b") { fromSq = mirrorSquare(fromSq); toSq = mirrorSquare(toSq); }
  const df = fileOf(toSq) - fileOf(fromSq);
  const dr = rankOf(toSq) - rankOf(fromSq);
  const plane = deltaToPlane(df, dr, promotion);
  return fromSq * 73 + plane;
}

File: trainer/chessai/encoding.py
"""Board <-> tensor encoding, and move <-> policy-index mapping.

Conventions
-----------
* The board is always encoded from the perspective of the side to move: if it is
  Black's turn we mirror the board vertically and swap colors, so the network
  always "sees" itself as White moving up the board. Moves are mirrored to match.
* Policy space uses the AlphaZero 8x8x73 = 4672 encoding:
    - 56 "queen-like" planes: 8 compass directions x 7 distances
    -  8 knight planes
    -  9 underpromotion planes: 3 move directions x {knight, bishop, rook}
  Queen promotions are encoded as ordinary forward/diagonal pawn pushes and
  decoded back to a queen promotion automatically.

The plane index for a from-square is `from_square * 73 + plane`, so the flat
policy vector has length 64 * 73 = 4672.
"""

from __future__ import annotations

import chess
import numpy as np

# ----------------------------------------------------------------------------
# Board planes
# ----------------------------------------------------------------------------
# 12 piece planes (own P,N,B,R,Q,K then opponent P,N,B,R,Q,K)
# + 4 castling rights (own short, own long, opp short, opp long)
# + 1 en-passant file marker
# + 1 fifty-move-clock (scaled)
N_PLANES = 18

_PIECE_ORDER = [
    chess.PAWN,
    chess.KNIGHT,
    chess.BISHOP,
    chess.ROOK,
    chess.QUEEN,
    chess.KING,
]


def board_to_planes(board: chess.Board) -> np.ndarray:
    """Return an (N_PLANES, 8, 8) float32 tensor from side-to-move perspective."""
    planes = np.zeros((N_PLANES, 8, 8), dtype=np.float32)
    stm = board.turn  # side to move
    mirror = stm == chess.BLACK

    for sq in chess.SQUARES:
        piece = board.piece_at(sq)
        if piece is None:
            continue
        idx_sq = chess.square_mirror(sq) if mirror else sq
        rank = chess.square_rank(idx_sq)
        file = chess.square_file(idx_sq)
        # own pieces in planes 0-5, opponent in 6-11
        own = piece.color == stm
        base = 0 if own else 6
        plane = base + _PIECE_ORDER.index(piece.piece_type)
        planes[plane, rank, file] = 1.0

    # castling rights, from side-to-move perspective
    own_k = board.has_kingside_castling_rights(stm)
    own_q = board.has_queenside_castling_rights(stm)
    opp_k = board.has_kingside_castling_rights(not stm)
    opp_q = board.has_queenside_castling_rights(not stm)
    if own_k:
        planes[12, :, :] = 1.0
    if own_q:
        planes[13, :, :] = 1.0
    if opp_k:
        planes[14, :, :] = 1.0
    if opp_q:
        planes[15, :, :] = 1.0

    # en passant target file
    if board.ep_square is not None:
        ep = chess.square_mirror(board.ep_square) if mirror else board.ep_square
        planes[16, :, chess.square_file(ep)] = 1.0

    # fifty-move clock, scaled to ~[0,1]
    planes[17, :, :] = min(board.halfmove_clock, 100) / 100.0

    return planes


# ----------------------------------------------------------------------------
# Move <-> policy index
# ----------------------------------------------------------------------------
N_MOVES = 64 * 73  # 4672

# 8 compass directions as (file_delta, rank_delta)
_DIRECTIONS = [
    (0, 1),    # N
    (1, 1),    # NE
    (1, 0),    # E
    (1, -1),   # SE
    (0, -1),   # S
    (-1, -1),  # SW
    (-1, 0),   # W
    (-1, 1),   # NW
]

_KNIGHT_DELTAS = [
    (1, 2), (2, 1), (2, -1), (1, -2),
    (-1, -2), (-2, -1), (-2, 1), (-1, 2),
]

# underpromotion piece order
_UNDERPROMO = [chess.KNIGHT, chess.BISHOP, chess.ROOK]


def _mirror_square(sq: int) -> int:
    return chess.square_mirror(sq)


def move_to_index(move: chess.Move, board: chess.Board) -> int:
    """Encode a legal move (in board's own frame) to a policy index.

    Mirrors the move when it is Black to move so it lives in the canonical
    side-to-move frame used by ``board_to_planes``.
    """
    mirror = board.turn == chess.BLACK
    from_sq = _mirror_square(move.from_square) if mirror else move.from_square
    to_sq = _mirror_square(move.to_square) if mirror else move.to_square

    ff, fr = chess.square_file(from_sq), chess.square_rank(from_sq)
    tf, tr = chess.square_file(to_sq), chess.square_rank(to_sq)
    df, dr = tf - ff, tr - fr

    plane = _delta_to_plane(df, dr, move.promotion)
    return from_sq * 73 + plane


def _delta_to_plane(df: int, dr: int, promotion) -> int:
    # underpromotions (knight/bishop/rook): always a pawn pushing "up" one rank
    if promotion is not None and promotion != chess.QUEEN:
        piece_idx = _UNDERPROMO.index(promotion)
        # df in {-1, 0, 1} -> capture-left, push, capture-right
        return 64 + piece_idx * 3 + (df + 1)

    # knight moves
    if (df, dr) in _KNIGHT_DELTAS:
        return 56 + _KNIGHT_DELTAS.index((df, dr))

    # queen-like moves (includes queen promotions, encoded as the push)
    dist = max(abs(df), abs(dr))
    step = (0 if df == 0 else df // abs(df), 0 if dr == 0 else dr // abs(dr))
    dir_idx = _DIRECTIONS.index(step)
    return dir_idx * 7 + (dist - 1)


def index_to_move(index: int, board: chess.Board) -> chess.Move | None:
    """Decode a policy index back into a concrete move in ``board``'s frame.

    Returns None if the decoded move is illegal in this position.
    """
    mirror = board.turn == chess.BLACK
    from_sq = index // 73
    plane = index % 73

    ff, fr = chess.square_file(from_sq), chess.square_rank(from_sq)

    promotion = None
    if plane < 56:
        dir_idx = plane // 7
        dist = (plane % 7) + 1
        df = _DIRECTIONS[dir_idx][0] * dist
        dr = _DIRECTIONS[dir_idx][1] * dist
    elif plane < 64:
        df, dr = _KNIGHT_DELTAS[plane - 56]
    else:
        u = plane - 64
        promotion = _UNDERPROMO[u // 3]
        df = (u % 3) - 1
        dr = 1  # underpromotion always advances one rank in canonical frame

    tf, tr = ff + df, fr + dr
    if not (0 <= tf < 8 and 0 <= tr < 8):
        return None
    to_sq = chess.square(tf, tr)

    # un-mirror back to the board's real frame
    if mirror:
        from_sq_real = _mirror_square(from_sq)
        to_sq_real = _mirror_square(to_sq)
    else:
        from_sq_real, to_sq_real = from_sq, to_sq

    # auto queen-promotion: pawn reaching last rank via a queen-like plane
    if promotion is None and plane < 56:
        piece = board.piece_at(from_sq_real)
        if piece is not None and piece.piece_type == chess.PAWN:
            last_rank = 7 if board.turn == chess.WHITE else 0
            if chess.square_rank(to_sq_real) == last_rank:
                promotion = chess.QUEEN

    move = chess.Move(from_sq_real, to_sq_real, promotion=promotion)
    if move in board.legal_moves:
        return move
    return None


def legal_move_indices(board: chess.Board) -> dict[int, chess.Move]:
    """Map every legal move in ``board`` to its policy index."""
    out: dict[int, chess.Move] = {}
    for mv in board.legal_moves:
        out[move_to_index(mv, board)] = mv
    return out
